src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs

```diff
@@ -301,6 +302,11 @@ internal class CodeActionEditHandlerService(
                 var navigationService = workspace.Services.GetRequiredService<IDocumentNavigationService>();
                 await navigationService.TryNavigateToPositionAsync(
                     this._threadingContext, workspace, navigationOperation.DocumentId, navigationOperation.Position, cancellationToken).ConfigureAwait(false);
+                if (navigationOperation is VSTypeScriptInlineRenameOperation)
+                {
+                    var openDocument = workspace.CurrentSolution.GetRequiredDocument(navigationOperation.DocumentId);
+                    _renameService.StartInlineSession(openDocument, new TextSpan(navigationOperation.Position, 0), cancellationToken);
```

Comment:
this would move into the Apply method of the new operation.

Reply:
I tried this approach but having the operation execute with all the other operations causes problems. Specifically, it looks like the undo transaction open while running the operation gives an exception: 
`System.Runtime.InteropServices.COMException: 'Catastrophic failure (Exception from HRESULT: 0x8000FFFF (E_UNEXPECTED))'`

which is coming from:
`Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.InlineRename.VisualStudioInlineRenameUndoManagerServiceFactory.InlineRenameUndoManager.GetUndoUnits(Microsoft.VisualStudio.OLE.Interop.IOleUndoManager undoManager)`

When I remove the transaction (actually just moved the commit before all the operations run), there is no error. So it seems like there's not a clean way of doing this without making at least some change to the CodeActionEditHandlerService.

Reply:
interesting.  and that's a fair point about a problem with the current system.  it presumes the apply calls will just make edits that can then be undone in the global undo transaction.

So an alternative i'd be ok with would be to make an internal, sealed `StartInlineRenameSessionOperation` class that is a sibling to DocumentNavigationOperation.  

we'd special case looking for that and starting the inline rename session.  You could then expose that operation in your EA layer with some sort of factory-method.

Reply:
@PranavSenthilnathan Can you link to the code where the transaction is created and completed? I'm specifically wondering what change you made here:
> When I remove the transaction (actually just moved the commit before all the operations run)

Reply:
It was moving https://github.com/dotnet/roslyn/blob/afabc2f93fb807375f6fdb3b8ee9dfdd0ac6c1dd/src/EditorFeatures/Core/CodeActions/CodeActionEditHandlerService.cs#L179 right above the `try` which is a non-sensical change but I just wanted to see if the transaction was causing the issue.